examples/tls_alpn_provision.rs

```diff
@@ -0,0 +1,363 @@
+//! Demonstrate using instant-acme, tokio-rustls and rcgen to provision a certificate with
+//! TLS-ALPN-01
+//!
+//! You can run this example on a server with :443 open to the internet by running:
+//!  cargo run \
+//!   --package instant-acme \
+//!   --example tls_alpn_provision -- \
+//!     --directory-url=https://acme-staging-v02.api.letsencrypt.org/directory \
+//!     --name foo.example.com
+//!
+//! Make sure your DNS has been set up to resolve 'foo.example.com' to the server's IP address.
+//! This will issue a certificate for 'foo.example.com' from the Let's Encrypt staging server.
+//!
+//! You can run this example with Pebble by running:
+//!   cargo run \
+//!     --package instant-acme \
+//!     --example tls_alpn_provision -- \
+//!       --directory-url=https://localhost:14000/dir \
+//!       --server-root-cert-path=<path to pebble CA cert> \
+//!       --port 5001 \
+//!       --name foo.example.com
+//!
+//! Make sure Pebble is running using its default ports and your DNS has been set up to resolve
+//! 'foo.example.com' to localhost (or use Pebble with a mock DNS server like pebble-challtestsrv).
+
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use clap::Parser;
+use hyper_util::client::legacy::Client as HyperClient;
+use hyper_util::rt::TokioExecutor;
+use rcgen::{CertificateParams, CustomExtension, DistinguishedName, KeyPair};
+use rustls::crypto::CryptoProvider;
+use rustls::server::{Acceptor, ClientHello, ResolvesServerCert};
+use rustls::sign::CertifiedKey;
+use rustls::{ClientConfig, RootCertStore, ServerConfig};
+use rustls_pki_types::pem::PemObject;
+use rustls_pki_types::{CertificateDer, PrivatePkcs8KeyDer};
+use rustls_platform_verifier::ConfigVerifierExt;
+use tokio::net::{TcpListener, TcpStream};
+use tokio::sync::mpsc;
+use tokio::time::sleep;
+use tokio::{select, task};
+use tokio_rustls::LazyConfigAcceptor;
+use tracing::{info, warn};
+
+use instant_acme::{
+    Account, AuthorizationStatus, ChallengeType, HttpClient, Identifier, LetsEncrypt, NewAccount,
+    NewOrder, OrderStatus,
+};
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    tracing_subscriber::fmt::init();
+
+    let opts = Options::parse();
+
+    // Build a Rustls config that can validate the ACME server's certificate.
+    let rustls_config = if let Some(roots_pem) = &opts.server_root_cert_path {
```

Comment:
Why have the `server_root_cert_path` option, is that for local testing? Seems overkill for the example.

Reply:
> Why have the server_root_cert_path option, is that for local testing?

Yeah exactly, it was handy to do the initial development/testing locally against Pebble. It might also be handy to allow running the example in CI (but didn't implement that in this branch).

I don't feel strongly if you prefer it left out.